doc/charm++/manual.rst

```diff
@@ -2894,96 +2927,89 @@ elements migratable (see migration section above) and choose a load
-at appropriate (generally at a pre-existing synchronization boundary) to
-trigger load balancing by inserting a function call (AtSync) in the
-application source code.
-
-In the prior two load balancing modes, users do not need to worry about
-how to start load balancing. However, in one scenario, those automatic
-load balancers will fail to work - when array elements are created by
-dynamic insertion. This is because the above two load balancing modes
-require an application to have fixed the number of objects at the time
-of load balancing. The array manager needs to maintain a head count of
-local array elements for the local barrier. In this case, the
-application must use the *manual mode* to trigger load balancer.
-
-The detailed APIs of these three methods are described as follows:
-
-#. **Periodical load balancing mode**: In the default setting, load
-   balancing happens whenever the array elements are ready, with an
-   interval of 1 second. It is desirable for the application to set a
-   larger interval using +LBPeriod runtime option. For example
-   “+LBPeriod 5.0” can be used to start load balancing roughly every 5
-   seconds. By default, array elements may be asked to migrate at any
-   time, provided that they are not in the middle of executing an entry
-   method. The array element’s variable usesAtSync being false
-   attributes to this default behavior.
-
-#. **At sync mode**: Using this method, elements can be migrated only at
-   certain points in the execution when the application invokes
-   AtSync(). In order to use the at sync mode, one should set usesAtSync
-   to true in the array element constructor. When an element is ready to
-   migrate, call AtSync()  [6]_. When all local elements call AtSync,
-   the load balancer is triggered. Once all migrations are completed,
-   the load balancer calls the virtual function
-   ArrayElement::ResumeFromSync() on each of the array elements. This
+There are two different ways to use load balancing for chare arrays to
+meet different needs of the applications. These methods are different
+in how and when a load balancing phase starts. The two methods are:
+**AtSync mode** and **periodic load balancing mode**.  In *AtSync
+mode*, the application invokes the load balancer explicitly at an
+appropriate location (generally at a pre-existing synchronization
+boundary) to trigger load balancing by inserting a function call
+(``AtSync()``) in the application source code.  In *periodic load
+balancing mode*, a user specifies only how often load balancing is to
+occur, using the *+LBPeriod* runtime option to specify the time
+interval.
+
+The detailed APIs of these two methods are described as follows:
+
+#. **AtSync mode**: Using this method, elements can be migrated only
+   at certain points in the execution, when the application invokes
+   ``AtSync()``. In order to use AtSync mode, one should set
+   ``usesAtSync`` to true in the array element constructor. When an
```

Comment:
The wording of this para is maybe a bit confused w.r.t. where usesAtSync should be set. Although I'm not exactly sure how to improve it. IMO it should state right away that usesAtSync should be set in the constructor of all elements which take part in the synchronization process. Those that do not set atsync will still be migrated though so their PUP should be able to handle migration at any time in their life cycle. Currently this is stated at the end of the para instead.

This also leaves out any description of AtSync and arrays which are bound together, which maybe is what it should do? Using both together has come up as a point of confusion though in both paratreet and Quinoa. Maybe it's best to discuss in the section about bound arrays though? Similarly there is no mention of how load is represented for bound objects (or any objects really, including those where setMigration is set to false). But this is probably more relevant for the section where we discuss how to write new strategies.

Reply:
I discussed this with Eric offline a bit, and I've expanded the description of AtSync and participation a bit to hopefully make it clearer.

I think the discussions about bound arrays and about how load is represented are both outside the scope of this section.